scripts/find_invalid_waivers.py

```diff
@@ -0,0 +1,185 @@
+
+
+def unwaive_note(waive_text, note):
+    return note.lstrip(waive_text).rstrip(')')
+
+
+def match_result_mark_waiver(regexes_matched_list, version, arch, status, name, note):
+    """This function is an updated version of the match_result() function from the lib/waive.py."""
+    v = version.split('.')
+
+    # make sure "'something' in name" always works
+    if name is None:
+        name = ''
+    if note is None or note == '[]':
+        note = ''
+
+    if '(waived pass)' in note:
+        note = unwaive_note('(waived pass)', note)
+    elif '(waived fail)' in note:
+        note = unwaive_note('(waived fail)', note)
+        status = 'fail'
+    elif '(waived error)' in note:
+        note = unwaive_note('(waived error)', note)
+        status = 'error'
+
+    objs = {
+        'status': status,
+        'name': name,
+        'note': note,
+        'arch': arch,
+        'rhel': _FakeRhel(v[0], v[1]),
+        'oscap': '',
+        'env': '',
+        'Match': waive.Match,
+    }
+
+    for section in _sections_cache:
+        if any(x.fullmatch(name) for x in section.regexes):
+            ret = eval(section.python_code, objs, None)
+
+            if not isinstance(ret, waive.Match):
+                if not isinstance(ret, bool):
+                    raise RuntimeError(f"waiver python code did not return bool or Match: {ret}")
+                ret = waive.Match(ret)
+
+            if ret:  # both regex and python code matched
+                add_regexes_to_list(regexes_matched_list, section.regexes, bool(status != 'pass'))
+
+
+def load_and_process_results(file, regexes_matched_list):
+    with gzip.open(file, 'rt') as f:
+        for line in f:
+            line = line.strip()
+            version, arch, status, name, note = line.split('\t')
+
+            if version == 'rhel':  # skip the header
+                continue
+            if status not in ['pass', 'fail', 'error', 'warn']:
+                continue
+
+            match_result_mark_waiver(regexes_matched_list, version, arch, status, name, note)
+
+
+def get_invalid_waivers(result_file_list):
+    # list of MatchedWaivers tuples
+    regexes_matched_list = []
+
+    global _sections_cache
+    _sections_cache = list(waive.collect_waivers())
+
+    for file in result_file_list:
+        load_and_process_results(file, regexes_matched_list)
+
+    print(
+        "===============================================================\n"
+        "The following waivers are no longer valid, they either did not\n"
+        "match any test results or only matched the 'pass' test results:\n"
+        "===============================================================\n",
+    )
+    for section in _sections_cache:
+        index = find_regexes_in_list(regexes_matched_list, section.regexes)
+        # if the section did not match any test results or only matched the 'pass' test results:
+        if index is None or regexes_matched_list[index].non_pass_result is False:
```

Comment:
You spend all the effort of recording separately whether the waiver matched + whether it didn't match on pass, and then you seem to use only one boolean state here.

Wouldn't it make more sense to throw away the True/False (whether a section matched a pass) and only record sections that didn't match pass in the first place?

IOW putting pass-matching ones into the same bucket as never-matching ones, without making a distinction?